targets/TARGET_NORDIC/TARGET_NRF5x/pwmout_api.c

```diff
@@ -33,379 +33,306 @@
-            m_pwm[i].signal.duty_us   = 100000;
-            m_pwm[i].signal.duty      = 0.5f;
-            
-            obj->pwm_struct  = &m_pwm[i];
-
-            internal_pwmout_exe(obj, true, true);
-
-            break;
-        }
-    }
-    
-    MBED_ASSERT(i != PWM_INSTANCE_COUNT); // assert if free instance was not found.
+    DEBUG_PRINTF("pwmout_init: %d\r\n", pin);
+
+    MBED_ASSERT(obj);
+
+    /* Get hardware instance from pinmap. */
+    int instance = pin_instance_pwm(pin);
+
+    MBED_ASSERT(instance < (int)(sizeof(nordic_nrf5_pwm_instance) / sizeof(nrf_drv_pwm_t)));
+
+    /* Populate PWM object with default values. */
+    obj->instance = instance;
+    obj->pin = pin;
+    obj->pulse = 0;
+    obj->period = MAX_PWM_COUNTERTOP;
+    obj->percent = 0;
+    obj->sequence.values.p_common = &obj->pulse;
+    obj->sequence.length = NRF_PWM_VALUES_LENGTH(obj->pulse);
+    obj->sequence.repeats = 0;
+    obj->sequence.end_delay = 0;
+
+    /* Initialize PWM instance. */
+    nordic_pwm_init(obj);
 }
 
+/** Deinitialize the pwmout object
+ *
+ * Parameter obj The pwmout object
+ */
 void pwmout_free(pwmout_t *obj)
 {
-    nrf_drv_pwm_uninit( (nrf_drv_pwm_t*) obj->pwm_struct );
+    DEBUG_PRINTF("pwmout_free\r\n");
+
+    MBED_ASSERT(obj);
 
-    m_pwm[obj->pwm_channel].p_pwm_driver = NULL;
+    /* Uninitialize PWM instance. */
+    nrf_drv_pwm_uninit(&nordic_nrf5_pwm_instance[obj->instance]);
 }
 
+/** Set the output duty-cycle in range <0.0f, 1.0f>
+ *
+ * pulse 0.0f represents 0 percentage, 1.0f represents 100 percent.
+ * Parameter obj     The pwmout object
+ * Parameter percent The floating-point percentage number
+ */
 void pwmout_write(pwmout_t *obj, float percent)
 {
-    
-    if (percent < 0)
-    {
-        percent = 0;
-    }
-    else if (percent > 1)
-    {
-        percent = 1;
-    }
-    
-    pwm_signal_t * p_pwm_signal = &(((pwm_t*)obj->pwm_struct)->signal);
-    
-    p_pwm_signal->duty = percent;
-    
-    int us  = (((int)p_pwm_signal->period_us) * percent);
-    
-    pwmout_pulsewidth_us(obj, us);
+    DEBUG_PRINTF("pwmout_write: %f\r\n", percent);
+
+    /* Find counts based on period. */
+    uint16_t pulse = obj->period * percent;
+
+    /* Ensure we don't overcount. */
+    obj->pulse = (pulse > MAX_PWM_COUNTERTOP) ? MAX_PWM_COUNTERTOP : pulse;
+
+    /* Store actual percentage passed as parameter to avoid floating point rounding errors. */
+    obj->percent = percent;
+
+    /* Set new duty-cycle. */
+    ret_code_t result = nrf_drv_pwm_simple_playback(&nordic_nrf5_pwm_instance[obj->instance],
+                                                    &obj->sequence,
+                                                    1,
+                                                    NRF_DRV_PWM_FLAG_LOOP);
+
+    MBED_ASSERT(result == NRF_SUCCESS);
 }
 
+/** Read the current float-point output duty-cycle
+ *
+ * Parameter obj The pwmout object
+ * Return A floating-point output duty-cycle
+ */
```

Comment:
why are we having these docs duplicates in the code file? they are defined in the header file

Reply:
I don't want to bounce back and forth between files to read comments. Besides, this is where the function is implemented. It's important to document what the function is doing and what the parameters are supposed to be.

Reply:
> It's important to document what the function is doing and what the parameters are supposed to be.

It is, in the header file - one place to have it all. This is duplication that is hard to maintain (in this case each HAL implementation would need an update when docs are updated in the header file) and easy to get out of the sync. I would not scatter these in any code file if the API is public.

Reply:
The comment serve a different purpose though. The description in `*_api.h` is the API definition, and like you said that is the golden standard. The description here is about the current implementation.

If the API changes in the future you will be able to see immediately that the API description and implementation description is different, which means the implementation maybe no longer follows the API and should be reviewed.